### PR TITLE
Add grouped conv bwd weight dl instances and new layout

### PR DESCRIPTION
**Pull request by bartekxk**

### CODE REVIEW

File: library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/device_grouped_conv2d_bwd_weight_dl_gnhwc_gkyxc_gnhwk_bf16_instance.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#include "ck/library/tensor_operation_instance/add_device_operation_instance.hpp"
+
```
Comment (bwroblew) [inline library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/device_grouped_conv2d_bwd_weight_dl_gnhwc_gkyxc_gnhwk_bf16_instance.cpp]: please remove blank line here and in other instance files (same place)

File: library/src/tensor_operation_instance/gpu/grouped_conv3d_bwd_weight/device_grouped_conv3d_bwd_weight_dl_gndhwc_gkzyxc_gndhwk_bf16_instance.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#include <cstdlib>
+
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
+#include "ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_dl.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
```
Comment (bwroblew) [inline library/src/tensor_operation_instance/gpu/grouped_conv3d_bwd_weight/device_grouped_conv3d_bwd_weight_dl_gndhwc_gkzyxc_gndhwk_bf16_instance.cpp]: looks unnecessary, huh?
same in other 3d instance files

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp [hunk 3]
```diff
@@ -27,28 +29,59 @@ class TestGroupedConvndBwdWeight : public ::testing::Test
     using NDimSpatial = std::tuple_element_t<6, Tuple>;
 
     std::vector<ck::utils::conv::ConvParam> conv_params;
-    ck::index_t split_k{2};
+    std::vector<ck::index_t> split_ks{1, 2};
+
+    bool skip_case(const ck::utils::conv::ConvParam& params, const ck::index_t split_k)
+    {
+        // K=1 or C=1 is supported only by dl kernel
```
Comment (bwroblew) [inline test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp]: do you mean cases with `K=1 or C=1` or cases `K or C odd`? The comment says something else than the code

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp [hunk 4]
```diff
@@ -27,28 +29,59 @@ class TestGroupedConvndBwdWeight : public ::testing::Test
...
+        // dl kernel is only supported for split_k=1
```
Comment (bwroblew) [inline test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp]: just to make it consistent, here and below
```suggestion
        // K=1 or C=1 is supported only by DL kernel
        // DL kernel is only supported for split_k=1
```

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp [hunk 5]
```diff
@@ -27,28 +29,59 @@ class TestGroupedConvndBwdWeight : public ::testing::Test
...
+        if constexpr(std::is_same_v<InDataType, ck::half_t>)
+        {
+            if(split_k != 1 && (params.K_ % 2 != 0 || params.C_ % 2 != 0))
+            {
+                return true;
+            }
+        }
+
+        // 1d nhwgc is only supported by dl kernel
+        // dl kernel is only supported for split_k=1
```
Comment (bwroblew) [inline test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp]:
```suggestion
        // 1D NHWGC is only supported by DL kernel
        // DL kernel is only supported for split_k=1
```

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp [hunk 6]
```diff
@@ -27,28 +29,59 @@ class TestGroupedConvndBwdWeight : public ::testing::Test
...
 
     std::vector<ck::utils::conv::ConvParam> conv_params;
-    ck::index_t split_k{2};
+    std::vector<ck::index_t> split_ks{1, 2};
+
+    bool skip_case(const ck::utils::conv::ConvParam& params, const ck::index_t split_k)
+    {
+        // K=1 or C=1 is supported only by dl kernel
+        // dl kernel is only supported for split_k=1
+        if constexpr(std::is_same_v<InDataType, ck::half_t>)
```
Comment (bwroblew) [inline test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:38]: If the datatype is sth else than fp16, is your comment still true? If not, maybe would be better to update it to not cause any confusion?

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp [hunk 7]
```diff
@@ -27,28 +29,59 @@ class TestGroupedConvndBwdWeight : public ::testing::Test
...
+        {
+            if(split_k != 1 && (params.K_ % 2 != 0 || params.C_ % 2 != 0))
+            {
+                return true;
+            }
+        }
+
+        // 1d nhwgc is only supported by dl kernel
+        // dl kernel is only supported for split_k=1
+        if constexpr(std::is_same_v<InLayout, NWGC> && std::is_same_v<OutLayout, NWGK>)
```
Comment (bwroblew) [inline test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:48]: `NWGC` or `NHWGC`?

Comment (bartekxk) [reply test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:48]: Fixed to NWGC.

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp (hunk 6 above)
Comment (bartekxk) [reply test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:38]: Fixed comment. It only applies for fp16

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp (hunk 5 above)
Comment (bartekxk) [reply test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp]: Fixed.

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp (hunk 4 above)
Comment (bartekxk) [reply test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp]: Fixed.

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp (hunk 3 above)
Comment (bartekxk) [reply test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp]: Fixed to K or C are odd.

File: library/src/tensor_operation_instance/gpu/grouped_conv3d_bwd_weight/device_grouped_conv3d_bwd_weight_dl_gndhwc_gkzyxc_gndhwk_bf16_instance.cpp (hunk 2 above)
Comment (bartekxk) [reply library/src/tensor_operation_instance/gpu/grouped_conv3d_bwd_weight/device_grouped_conv3d_bwd_weight_dl_gndhwc_gkzyxc_gndhwk_bf16_instance.cpp]: Removed.

File: library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/device_grouped_conv2d_bwd_weight_dl_gnhwc_gkyxc_gnhwk_bf16_instance.cpp (hunk 1 above)
Comment (bartekxk) [reply library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/device_grouped_conv2d_bwd_weight_dl_gnhwc_gkyxc_gnhwk_bf16_instance.cpp]: Removed.

**Review by bwroblew — APPROVED** [review]

File: library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_dl_instance.hpp [hunk 8]
```diff
@@ -55,8 +55,8 @@ using device_grouped_conv2d_fwd_dl_f16_instances = std::tuple<
            // ########################################|            |       |        |            |        |        |         |          |          |          |            |             |              |               |                  |      |      |      |      |   |           |           |       |              |              |                   |                     |               |               |                    |                   |                    |                   |                     |               |               |                    |                   |                    |                    |                |                   |
         // generic instances
         // TODO: Change to ScalarPerVector = 1 when inner_product<half_t, half_t, float> will be supported
```
Comment (aosewski) [inline library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_dl_instance.hpp:57]: You can now remove this comment.

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp [hunk 9]
```diff
@@ -27,28 +29,59 @@ class TestGroupedConvndBwdWeight : public ::testing::Test
     using NDimSpatial = std::tuple_element_t<6, Tuple>;
 
     std::vector<ck::utils::conv::ConvParam> conv_params;
-    ck::index_t split_k{2};
+    std::vector<ck::index_t> split_ks{1, 2};
+
+    bool skip_case(const ck::utils::conv::ConvParam& params, const ck::index_t split_k)
+    {
+        // K or C are odd is supported only by DL kernel (only applies to fp16)
+        // DL kernel is only supported for split_k=1
```
Comment (aosewski) [inline test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:37]:
```suggestion
        // Odd K or C values are supported only by DL kernel (only applies to fp16)
        // DL kernel supports currently supports only `split_k=1`
```

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp [hunk 10]
```diff
@@ -27,28 +29,59 @@ class TestGroupedConvndBwdWeight : public ::testing::Test
...
+        if constexpr(std::is_same_v<InDataType, ck::half_t>)
+        {
+            if(split_k != 1 && (params.K_ % 2 != 0 || params.C_ % 2 != 0))
+            {
+                return true;
+            }
+        }
+
+        // 1d NWGC is only supported by DL kernel
+        // DL kernel is only supported for split_k=1
```
Comment (aosewski) [inline test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:47]:
```suggestion
        // DL kernels support only 1D (NWGC) data layout and split_k=1
```

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_dl.hpp [hunk 11]
```diff
@@ -96,9 +100,27 @@ __global__ void
...
+    ignore = batch_count;
+    ignore = a_grid_desc_kbatch_k0_m0_m1_k1;
+    ignore = b_grid_desc_kbatch_k0_n0_n1_k1;
+    ignore = c_grid_desc_m0_m10_m11_n0_n10_n11;
+    ignore = block_2_ctile_map;
+    ignore = compute_ptr_offset_of_batch;
+
+    compute_ptr_offset_of_batch.GetAPtrOffset(0);
+    compute_ptr_offset_of_batch.GetBPtrOffset(0);
+    compute_ptr_offset_of_batch.GetCPtrOffset(0);
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_dl.hpp:116]: Is it needed if when you have line 112 ?

**Review by aosewski — APPROVED** [review]
Just have few small comments.

**Comment by zjing14** [comment]
@bartekxk Please create a post-merge PR for Adam's comments.

File: library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_dl_instance.hpp (hunk 8 above)
Comment (bartekxk) [reply library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_dl_instance.hpp:57]: Done.

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp (hunk 9 above)
Comment (bartekxk) [reply test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:37]: Done.

File: test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp (hunk 10 above)
Comment (bartekxk) [reply test/grouped_convnd_bwd_weight/test_grouped_convnd_bwd_weight.cpp:47]: Your version is not always true. Dl kernels supports also other num dimensions and layouts. I prefer to stay with my version.

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_dl.hpp (hunk 11 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_dl.hpp:116]: Yes to avoid unused methods.

Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_dl.hpp:116]: Removed.

**Pull request merged**